libs/vkd3d/device.c

```diff
@@ -4400,18 +4401,142 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreateSharedHandle(d3d12_device_if
+    }
+
+    FIXME("Creating shared handle for type of object %p unsupported\n", object);
     return E_NOTIMPL;
 }
 
+static inline bool handle_is_kmt_style(HANDLE handle)
+{
+    return ((ULONG_PTR)handle & 0x40000000) && ((ULONG_PTR)handle - 2) % 4 == 0;
+}
+
 static HRESULT STDMETHODCALLTYPE d3d12_device_OpenSharedHandle(d3d12_device_iface *iface,
         HANDLE handle, REFIID riid, void **object)
 {
-    FIXME("iface %p, handle %p, riid %s, object %p stub!\n",
+    struct d3d12_device *device = impl_from_ID3D12Device(iface);
+    HRESULT hr;
+
+    TRACE("iface %p, handle %p, riid %s, object %p\n",
             iface, handle, debugstr_guid(riid), object);
 
+    if (IsEqualGUID(riid, &IID_ID3D12Resource))
+    {
+        struct DxvkSharedTextureMetadata metadata;
+        D3D12_HEAP_PROPERTIES heap_props;
+        struct d3d12_resource *resource;
+        D3D12_RESOURCE_DESC1 desc;
+        bool kmt_handle = false;
+
+        if (handle_is_kmt_style(handle))
+        {
+            handle = vkd3d_open_kmt_handle(handle);
+            kmt_handle = true;
+
+            if (handle == INVALID_HANDLE_VALUE)
+            {
+                WARN("Failed to open KMT-style ID3D12Resource shared handle.\n");
+                *object = NULL;
+                return E_INVALIDARG;
+            }
+        }
+
+        if (!vkd3d_get_shared_metadata(handle, &metadata, sizeof(metadata), NULL))
+        {
+            WARN("Failed to get ID3D12Resource shared handle metadata.\n");
+            if (kmt_handle)
+                CloseHandle(handle);
+
+            *object = NULL;
+            return E_INVALIDARG;
+        }
+
+        desc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+        desc.Alignment = 0;
+        desc.Width = metadata.Width;
+        desc.Height = metadata.Height;
+        desc.DepthOrArraySize = metadata.ArraySize;
+        desc.MipLevels = metadata.MipLevels;
+        desc.Format = metadata.Format;
+        desc.SampleDesc = metadata.SampleDesc;
+        switch (metadata.TextureLayout)
+        {
+            case D3D11_TEXTURE_LAYOUT_UNDEFINED: desc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN; break;
+            case D3D11_TEXTURE_LAYOUT_ROW_MAJOR: desc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR; break;
+            case D3D11_TEXTURE_LAYOUT_64K_STANDARD_SWIZZLE: desc.Layout = D3D12_TEXTURE_LAYOUT_64KB_STANDARD_SWIZZLE; break;
+            default: desc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN;
+        }
+        desc.Flags = D3D12_RESOURCE_FLAG_ALLOW_SIMULTANEOUS_ACCESS;
+        if (metadata.BindFlags & D3D11_BIND_RENDER_TARGET)
+            desc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
+        if (metadata.BindFlags & D3D11_BIND_DEPTH_STENCIL)
+            desc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL;
+        if (metadata.BindFlags & D3D11_BIND_UNORDERED_ACCESS)
+            desc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
+        if (!(metadata.BindFlags & D3D11_BIND_SHADER_RESOURCE))
```

Comment:
DENY_SHADER_RESOURCE is only allowed for ALLOW_DEPTH_STENCIL.

Reply:
Ah okay, removed the check entirely then.  Should I instead only set DENY_SHADER_RESOURCE when it's a depth stencil?